### ISSUE TITLE
[Bug]: Speculative decoding (mtp and dflash) silently disables prefix-cache hits for hybrid GDN models on nightly; worked on v0.24.0

### ISSUE THREAD

**Issue by cgoudie** [vllm-project/vllm#54360]
(analysis by Fable 5)
### Your current environment

- DGX Spark (GB10, `sm_121`/`sm_121a`, aarch64, 128 GB unified memory), Docker `--network host --gpus all --ipc host`
- Image: `vllm/vllm-openai:nightly` → reports `v0.28.1rc1.dev43+g6f7df92a8`
- Comparison baseline on the same host: `vllm/vllm-openai:v0.24.0`
- Model: `unsloth/Qwen3.8-27B-NVFP4` (hybrid GDN: 48 linear-attention + 16 full-attention layers, built-in MTP head). Also reproduced with `RadixArk/Qwen3.8-27B-NVFP4` + `incoai/Qwen3.8-27B-DFlash2` (rev `dedf8df`).
- `--kv-cache-dtype fp8`, `--max-model-len 65536`, `--gpu-memory-utilization 0.30`, env `CUTE_DSL_ARCH=sm_121a`

### 🐛 Describe the bug

On the current nightly, enabling **any** speculative decoding method (built-in `mtp` or `dflash`) on a hybrid GDN model silently drops `vllm:prefix_cache_hits_total` to **0** — align-mode prefix caching engages at startup (log: `Mamba cache mode is set to 'align' for Qwen3_5ForConditionalGeneration`) but never hits, and repeated long prompts pay full re-prefill. The identical command minus `--speculative-config` hits normally. No crash, no warning — the caching is just inert, so this is easy to miss in production.

A/B on the same host, same model, same 18,219-token prompt sent twice (`max_tokens=10`):

| nightly config | TTFT req1 → req2 | `prefix_cache_hits_total` delta |
|---|---|---|
| APC only (no spec) | 9.46 s → **0.64 s** | **17,248** |
| APC + `{"method":"mtp","num_speculative_tokens":3}` | 23.6 s → 13.9 s* | **0** |
| APC + `{"method":"dflash", ..., "num_speculative_tokens":7}` | 12.9 s → 12.5 s | **0** |

\* the MTP row also shows a large decode regression on this platform (10.5 tok/s vs 19.7 on v0.24.0 with the same flags — possibly a separate sm_121 issue; noting it for context, not the subject of this report).

**On `v0.24.0`, MTP + align-mode APC coexist correctly** on the same host/model: repeated 9,901-token prompt → 8,000 tokens hit, TTFT 7.27 s → 1.33 s, with `mtp` k=3 active. So spec+APC coexistence for hybrids existed and regressed somewhere between v0.24.0 and the current nightly.

Repro (nightly):

```bash
docker run --rm --init --network host --gpus all --ipc host \
  -v /opt/models:/models -e CUTE_DSL_ARCH=sm_121a \
  vllm/vllm-openai:nightly /models/qwen3.8-27b-nvfp4 \
  --served-model-name bench --port 10099 \
  --max-model-len 65536 --gpu-memory-utilization 0.30 \
  --max-num-seqs 1 --kv-cache-dtype fp8 \
  --enable-prefix-caching \
  --speculative-config '{"method":"mtp","num_speculative_tokens":3}'
# send any >3,300-token prompt twice; watch vllm:prefix_cache_hits_total stay 0
# drop --speculative-config and the second send hits ~all full blocks
```

Possibly related but distinct: #53670 (hits occur but recompute per hit), #41884 (DFlash+APC IndexError crash — we see silent zero-hit, not a crash), #26201 (tracking).

Is the suppression intentional (spec-decode + hybrid APC support not yet wired on the new path), or a regression? If intentional, a startup warning like the align-mode one would save users from silently losing their cache. Happy to run further diagnostics on this hardware.

**Comment by jschmied** [comment]
Datapoint that may narrow the bisect window: **on the same model and the same hardware, but on
0.27.1-era builds rather than nightly, speculation reduces prefix-cache hits but does not zero
them.**

GB10, `unsloth/Qwen3.8-27B-NVFP4` with an FP8 head, measured 2026-08-26 with an agent-loop probe
(~7.3k seed, 6 turns, each appending the reply plus a new question), cold cache per arm, hits
from `prefix_cache_hits_total` deltas:

| arm | first hit turn | plateau | hit rate | mean TTFT (turns 2+) |
| --- | --- | --- | --- | --- |
| no spec | 2 | 6272 = 4 x 1568 | 69.4% | 0.66 s |
| **mtp K=3** | 3 | **4800 = 3 x 1600** | **42.5%** | 2.04 s |
| **dflash2 K=7** | 3 | **4944 = 3 x 1648** | **43.3%** | 1.93 s |

So on those builds the failure mode is *one whole block of reachable prefix lost* (4 blocks → 3)
plus one extra cold turn — not inert caching. MTP and DFlash2 land identically, so it is the
last-block drop rather than anything drafter-specific.

Versions, since that is the whole point of the datapoint: the MTP arm ran on 0.27.1 and the
DFlash2 arm on a build of the #52816 branch. **Neither is nightly**, and neither is v0.24.0. If
your nightly really reads 0 while 0.27.1 reads 42.5% on the same model, the regression landed
between those two, which is a much smaller window than nightly-vs-v0.24.0.

Two things from our measurements that might save you time:

**The block size tracks speculative depth** — 1568 at K=0, 1600 at K=3, 1648 at K=7. Any
cross-K comparison of raw hit *counts* is comparing different geometries, so it is worth
confirming the 0 is genuinely zero hits rather than a plateau that moved.

**`usage.cached_tokens` is inert** and reads 0 even on real hits, so it cannot be used to
cross-check `prefix_cache_hits_total`. You are already using the right counter; flagging it only
because it has caught us before and it fabricates exactly the symptom under discussion.

Happy to re-run the same probe on nightly on our GB10 if a second confirmation would help — say
so and I will post the arms.

**Comment by jschmied** [comment]
There is an open PR for what looks like exactly this, and it is not linked from here yet:
**#52244 — "[Bugfix][V1] Restore hybrid GDN prefix-cache hits under MTP spec decoding"**.

Its measured table is the tell. On Qwen3.5-122B-A10B with MTP, a 1072-token GDN page and
`--prefix-match-unit 67`:

| prompt | cached on replay | ceiling |
| --- | --- | --- |
| 1072 | **0** | 938 |
| 2000 | **0** | 1876 |
| 2144 | **0** | 2010 |
| 3000 | 1072 | 2881 |
| 3500 | 2144 | 3417 |

**Zero for some prompt lengths, partial for others** — from three rules composing badly: the
lookup caps at `prompt_len - 1` so the last token can produce logits, the MTP/EAGLE drafter
rewinds the hit one hash unit further, and the producer publishes the GDN state at the prompt's
own tail boundary.

**That reconciles your result with mine.** I posted 42.5% / 43.3% hit rates above and framed it as
a version difference; #52244 suggests it may be prompt geometry instead. Your probe replays one
fixed 18,219-token prompt, so if that length lands on a hash-unit boundary you would see 0 every
time. My agent-loop probe grows the prompt each turn (~7.3k seed, six turns, each appending a
reply plus a new question), so it lands on boundaries only sometimes — which is what a
partial-hit plateau looks like. Same defect, different sampling of it, and no version regression
required to explain either.

Cheap way to separate the two hypotheses on your setup: replay the same prompt at a handful of
lengths near your current one — say ±1, ±17, ±67 tokens — on the same nightly. If the zero moves
to a nonzero as the length shifts off the boundary, it is #52244 and not a nightly-vs-v0.24.0
regression. If it stays at 0 for every length, the bisect argument stands and #52244 is a
different bug that happens to look alike.

I have a GB10 and can run that sweep if it is useful — say the word.

**Comment by erdholion** [comment]
Corroborating datapoint on **v0.27.1** (partial-loss tier, matching @jschmied's numbers rather than nightly's zero):

Stack: Qwen3.8-27B (hybrid GDN), NVFP4 weights, fp8 e4m3 KV, FLASHINFER backend, RTX 5090, MTP k=3, `--no-async-scheduling`, prefix caching on.

Exact-repeat probe (same 7172-token prompt sent twice, greedy):
- req 1: `prefix_cache_queries` +7172, hits +0 (cold, expected)
- req 2: queries +7172, **hits +4800 → 66.9%** on a byte-identical prompt — ~2.4k tokens re-prefilled on every cached turn.
- Also confirming `usage.prompt_tokens_details.cached_tokens` is absent on this path, so client-side detection of the loss is not possible.

Caveat: this box is a production server, so I don't have a no-MTP control arm yet — but an exact repeat should sit near full-minus-alignment, not 67%. Happy to run the ±1/±17/±67 prompt-length boundary probe from #52244 on this config if that helps narrow the geometry.

**Comment by erdholion** [comment]
Follow-up: completed the controlled A/B on the same box (v0.27.1, Qwen3.8-27B hybrid GDN, NVFP4, fp8 e4m3 KV, FLASHINFER, RTX 5090). Same probe both arms: byte-identical prompt sent twice with a unique cold prefix per pair, exact token counts via `/tokenize`, greedy. `req2_hits` = prefix_cache hits delta on the repeat.

**No spec (control):**

| prompt_toks | req2_hits | hit rate |
|---|---|---|
| 1201 | 0 | 0% |
| 2401 | 1568 | 65.3% |
| 3601 | 3136 | 87.1% |
| 4801 | 4704 | 98.0% |
| 6001 | 4704 | 78.4% |
| 7201 | 6272 | 87.1% |

Law: `hits = floor((L-1)/1568) * 1568` — clean quantization to a **1568-token hash unit, zero units lost**.

**MTP k=3:**

| prompt_toks | req2_hits | hit rate |
|---|---|---|
| 1201 | 0 | 0% |
| 2401 | 0 | 0% |
| 3601 | 1600 | 44.4% |
| 4801 | 3200 | 66.7% |
| 6001 | 3200 | 53.3% |
| 7201 | 4800 | 66.7% |

Law: `hits = max(0, floor((L-1)/1600) - 1) * 1600` — the hash unit **grows to 1600** and **exactly one unit is always lost**. Boundary is razor-sharp (±1/±17/±67 probes around 4800: prompt 4800 → 1600 hits, prompt 4801 → 3200; the ±17/±67 points sit flat on the law).

Two observations that may help #52244:
1. The hash unit itself changes with spec: 1568 = 49×32 vs 1600 = 50×32 (block_size 32 here) — consistent with the mamba-align state-save interval being recomputed to cover the spec lookahead, which then misaligns every previously-cached boundary.
2. The "one lost unit" is exact and unconditional in our data — matching @jschmied's "one reachable prefix block lost" on 0.27.1.

Practical impact for agent workloads on this config: the zero-reuse threshold moves from 1,569 to 3,201 prompt tokens, and every cached turn re-prefills ~1.5–3.1k tokens that the no-spec arm reuses.

**Comment by chiptoe-svg** [comment]
Two data points that may split this into two bugs, from a DGX Spark GB10 (SM121) running `nvidia/Qwen3.6-35B-A3B-NVFP4` (hybrid GDN MoE), fp8 KV, FLASHINFER, MTP k=3 (`"moe_backend":"triton"` for the draft), prefix caching on, `--async-scheduling` on, `--max-model-len 65536`.

**1. The "partial-loss" tier reproduces exactly on tagged releases, and it is the same law @erdholion derived.** Exact-repeat probe (byte-identical prompt twice, unique cold prefix per pair, `max_tokens 4`, greedy; hits = Δ`prefix_cache_hits_total` on the repeat), five prompt lengths, two pairs each, on **v0.28.0**:

| prompt_tokens | repeat hits | hit rate |
|---|---|---|
| 3784 | 0 | 0% |
| 7542 | 4288 | 56.9% |
| 11292 | 8576 | 75.9% |
| 15019 | 12864 | 85.7% |
| 18770 | 15008 | 80.0% |

Fits `hits = max(0, floor((L−1)/2144) − 1) × 2144` at every row — so this model's align unit is **2144 tokens**, and MTP costs **exactly one unit**, the same "minus one unit" law as your Qwen3.8-27B/1600 numbers on 0.27.1. **v0.26.0 gives the identical 8576 at 11.3K** (measured on the same box earlier today), so within 0.26.0→0.28.0 nothing changed here. We can't run the no-spec control without taking production down, so I can't independently confirm "zero units lost" without MTP — your control covers that.

**2. Both of our builds ran Model Runner V1; the "zero hits" tier in the OP was on a nightly, which runs Model Runner V2.** The tagged releases here log no V2 line; `vllm/vllm-openai:nightly` (and eugr's `nightly-20260905`, which we also booted) logs `[gpu_worker.py:436] Using V2 Model Runner` — MRV2 became the default in #53183. There is a family of MRV2 draft-config regressions being fixed right now: #54826 (draft `attention_backend`, merged 09-04) and #54788 (draft `moe_backend`, approved, CI running). If prefix hits going to *zero* only on nightly is another "draft loaded with the wrong config" case — e.g. the draft's KV-cache/prefix hashing not seeing the align-mode setting — it would be worth bisecting against #53183 first, before looking inside the hybrid cache code. That would make this two issues: a one-unit tax on V1 that has existed since at least 0.26.0, and a total loss on V2.

Probe script is `prefix_probe.py` in https://gist.github.com/chiptoe-svg/fdc3d3d647108fcbe4d775408cb3281a if a like-for-like comparison is useful (`--tokens 4000,8000,12000,16000,20000 --pairs 2`).

**Comment by nickchan0412** [comment]
Root cause located on nightly `0.3.1.dev3+g0bfc7a15d` (2026-09-17), GB10 (SM121), and it is an **insertion-side** failure, not a lookup-side one:

1. `_annotate_eagle_groups()` cannot identify a draft group (plain MTP shares the target KV; there is no draft group by construction) and logs the "no KV cache group could be identified" warning from `_warn_if_unannotated_eagle_mamba`.
2. `KVCacheCoordinator.__init__` then takes the conservative fallback (`if use_eagle and not self.eagle_group_ids: self.eagle_group_ids = set(range(len(groups)))`) and flags **all** groups as eagle.
3. With eagle flagged, `get_replay_boundaries()` switches to its eagle branch: for a 4,137-token prompt with scheduler block 3200, both `resend` and `extension` floor to 3200, so the returned boundaries collapse to `{max(3200-3200,0), max(3200-3200,0)} = {0}`.
4. `cache_blocks()` therefore registers nothing above position 0 — the radix tree stays empty, so every query misses. Hits are zero because nothing was ever inserted.

**Mechanism-level workaround, verified on GB10 with a 125B hybrid GDN NVFP4 model (MTP k=3, align mode, 262K ctx):** put `"disable_eagle_block_drop": true` inside the speculative config. Scheduler then passes `use_eagle=False` to the coordinator, the fallback never fires, and caching returns to normal — measured **0 hits → 50.1% hits** (repeated 4,137-token prompt TTFT 3.9 s → 1.6 s) with MTP acceptance unchanged-to-better (44.5% → 49.4% with reduced-vocab drafting). Cost: the trailing-block drop protection is also lost (acceptable here; no acceptance regression observed).

Two related notes:
- This also makes `enable_mamba_fine_grained_prefix_cache` unreachable for group-less-MTP models: it gates on `eagle_group_ids` being non-empty (KVCacheManager init), which is exactly what the workaround empties. The feature needs to key off mamba align mode / partial-hash-hit machinery rather than eagle group membership.
- Corroborates #57032 (dflash drafter group never identified) — same conservative fallback, different method.

Suggestion: when the speculative method has no separate draft KV (plain MTP), treat it as "no eagle groups" instead of flagging every group, and gate the trailing-block drop on a draft KV group actually existing.

Assisted-by: AI agent (ZCode); analysis and all measurements reviewed and confirmed on hardware by the submitter.
